contracts/Endian.sol

```diff
@@ -0,0 +1,14 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.23;
+
+contract Endian {
+  /// reverses byte order to allow conversion between little endian and big
+  /// endian integers
+  function _byteSwap(bytes32 input) internal pure returns (bytes32 output) {
```

Comment:
Wouldn't it be much faster to do the conversion outside the contract?

Reply:
It would (and be cheaper as well), but the data that are converted with this are gathered from the contract's data and are not "input to function call", so IMHO not really possible without removing the goal that we don't want to submit the public inputs to the `submitProof()` call. Have a look here: https://github.com/codex-storage/codex-contracts-eth/pull/79/files#diff-5973808e32384782b5978a63e85b893b67d634be8aa42001f0ef480f189a0688R162-R193

Reply:
But thinking about it a bit more, maybe this conversion would in itself warrant actually dropping that goal in order to optimize it from a gas perspective. I don't know what is cheaper though. IMHO we don't need to analyze it now, but it would be good to have a look on it later on with some analysis.

Reply:
We cannot allow hosts to supply their own public inputs, that would defeat the security of the scheme. They have to come from values that are on-chain. 

What we could do for the two values that are converted is the following:
- for the merkle root in the storage request, we can change its definition from a bytes32 to a uint256, which gets rid of the need to interpret bytes as an integer (be it little or big endian)
- for the randomness, we could interpret that as a big-endian integer, instead of a little-endian integer in the nim code

This however has two disadvantages:
- when we move to sha256 instead of poseidon2, we have to change the type of the merkle root back to bytes32 again
- we get a scheme for the zero-knowledge proof where most of the time we interpret bytes as little-endian, but in a special case we interpret bytes as big-endian

Reply:
Yeah, I did not really though it through. I thought that "we would do assertions on the submitted public inputs" to verify if it is not "defeating the security", but that would only circle back to the need to the conversions 😅 Never mind.

Reply:
There isn't any particular reason we're using little endian in the client side, why not change it there?

Reply:
In fact, it was probably an oversight that we didn't set off with bigendian from the get go, given that the EVM uses bigendian.

Reply:
We chose little endian because the circom and arkworks tooling uses it.